### DISCUSSION TITLE
Add Support for Cascade Layers

### DISCUSSION THREAD

**Discussion by kyle-jorve** [vercel/next.js#43525] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

 Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 16.17.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.0.5
      eslint-config-next: 13.0.4
      react: 18.2.0
      react-dom: 18.2.0

### What browser are you using? (if relevant)

Chrome Version 107.0.5304.107

### Describe the Bug

I've created an _index.css file which I am importing into _app.tsx, like so: `import "../styles/base/_index.css";`

In _index.css I'm importing stylesheets into layers, like so:
```
@layer reset, theme, icons, typography, buttons, layout;

@import url("./normalize.css") layer(reset);
@import url("./theme.css") layer(theme);
@import url("./icons.css") layer(icons);
@import url("./typography.css") layer(typography);
@import url("./buttons.css") layer(buttons);
@import url("./layout.css") layer(layout);
```

However, when I run the dev environment, this is the result:
![image](https://user-images.githubusercontent.com/34990609/204375869-946bb1f8-426b-4679-94de-ba62bf583e2c.png)

This leads me to believe that Next does not yet support cascade layers.

### Expected Behavior

A cascade layer shouldn't be added to the media attribute of a style tag. Rather, the styles themselves should be wrapped in the layer declaration. Example:
```
@layer reset {
/* all reset styles here */
...
}
```

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/kyle-jorve/portfolio-site-next

### To Reproduce

Run `npm run dev`. Inspect the stylesheets in the `<head>`.

**Comment by philwolstenholme** [comment]
https://www.npmjs.com/package/postcss-import will help you out here, but you might run into https://github.com/vercel/next.js/issues/47585 depending on your setup

**Comment by PixeledCode** [comment]
@kyle-jorve were you able to resolve this issue?

**Comment by kyle-jorve** [comment]
@PixeledCode I went with a different method of using cascade layers, so I haven't tried anything new. Maybe this is solved in newer versions of Next, but I don't know. 🤷‍♀️

**Reply by PixeledCode** [reply]
Not solved. I'm still facing issue with imported cascade layer is not working. 
What other way you went with?

**Reply by kyle-jorve** [reply]
What I did was put styles into cascade layers inside each stylesheet, then import them into a single stylesheet like this:
```
@import url("./normalize.css");
@import url("./theme.css");
@import url("./icons.css");
...
```

This way the layer order is established by the order of the imports. It's not ideal, but it works.

**Comment by SkyGopnik** [comment]
`postcss.config.js`

```
module.exports = {
  plugins: [
    [
      "postcss-header",
      {
        header: "@layer reset, component, page;"
      }
    ]
  ]
};
```

Next.js version: 15.2.1-canary.5

This fix @layer priority for me, because next build add global css to the end of file.

**Comment by sentience** [comment]
Issue here: https://github.com/vercel/next.js/issues/55763

**Comment by sentience** [comment]
With the release of Tailwind v4, this is now more urgent as the ecosystem upgrades to this version that makes heavy use of CSS Cascade Layers.

In our projects, our apps typically import styles from some dependencies, then generate Tailwind styles that we often use to override those dependencies' styles. With Tailwind v3, CSS source order permitted this:

```css
@import url(~package/styles.css);

/* Style generated here can override package styles (above) thanks to source order */
@tailwind base;
@tailwind components;
@tailwind utilities;
```

With Tailwind v4, however, Tailwind-generated styles are all wrapped in CSS layers (e.g. `@layer base { /* styles */ }`), which means they will be applied _before_ any un-layered styles in our dependencies:

```css
@import url(~package/styles.css);

/* These styles are now layered, so cannot override package styles (above) */
@import "tailwindcss";
```

A straightforward fix for this would be to specify the layer to import dependency styles into:

```css
@layer theme, base, components, utilities;

@import url(~package/styles.css) layer(components);

/* Styles in the 'utilities' layer will now override package styles (above) */
@import "tailwindcss";
```

…but this is not supported in Next.js. So the only work-around is, as @kyle-jorve noted, is to update the dependencies' styles to wrap them in `@layer components { … }`, which isn't always practical.
